// @web-analytics/vue/lib/index.mjs
/**
 * name: @web-analytics/vue
 * version: v0.2.2
 * description: Website pageview analytics tool for Vue.js (Including VuePress , VitePress etc.) and multi-analytics-platform support.
 * author: chengpeiquan <chengpeiquan@chengpeiquan.com>
 * homepage: https://analytics.js.org/vue/
 * license: MIT
 */
var g = Object.defineProperty;
var w = (t, e, r) => e in t ? g(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r;
var c = (t, e, r) => (w(t, typeof e != "symbol" ? e + "" : e, r), r);
import { isBrowser as l } from "@bassist/utils";
import { SUPPORTED_ANALYTICS_PLATFORMS as f, createCnzzAnalytics as z, createBaiduAnalytics as P } from "@web-analytics/core";
class p {
  constructor({ platform: e, websiteIds: r, debug: n }) {
    c(this, "pluginId");
    c(this, "platform");
    c(this, "websiteIds");
    c(this, "instances");
    c(this, "debug");
    this.pluginId = "web-analytics-vue", this.platform = e, this.websiteIds = r || [], this.instances = [], this.debug = n ?? !1, this.init();
  }
  init() {
    try {
      switch (this.platform) {
        case "baidu": {
          this.websiteIds.forEach((e) => {
            const r = P({
              pluginId: this.pluginId,
              websiteId: e,
              debug: this.debug
            });
            this.instances.push(r);
          });
          break;
        }
        case "cnzz": {
          this.websiteIds.forEach((e) => {
            const r = z({
              pluginId: this.pluginId,
              websiteId: e,
              debug: this.debug
            });
            this.instances.push(r);
          });
          break;
        }
        default: {
          const e = f.join(", ");
          this.throwError(
            `Unsupported platform options, only supported: ${e}.`
          );
          break;
        }
      }
    } catch (e) {
      this.throwError(`Plugin initialization failed: ${e}.`);
    }
  }
  trackPageview(e) {
    this.instances.forEach((r) => r.trackPageview({ ...e }));
  }
  trackEvent(e) {
    this.instances.forEach((r) => r.trackEvent({ ...e }));
  }
  throwError(e) {
    throw new Error(`[${this.pluginId}] ${e}`);
  }
}
function d({ version: t }) {
  try {
    const [r] = t.split(".");
    return r ? Number(r) : 3;
  } catch {
    return 3;
  }
}
function h(t) {
  return f.includes(t) ? `$${t}Analytics` : "";
}
function y({
  analytics: t,
  platform: e
}) {
  return (r, n) => {
    if (!l)
      return;
    const s = new p({ platform: e, ...n });
    for (const i in s)
      Object.prototype.hasOwnProperty.call(s, i) && (t[i] = s[i]);
    const o = Object.getPrototypeOf(s);
    Object.setPrototypeOf(t, o);
    const a = h(e);
    if (a)
      switch (d(r)) {
        case 2: {
          r.prototype[a] = t;
          break;
        }
        case 3: {
          r.config.globalProperties[a] = t;
          break;
        }
      }
  };
}
function O() {
  const t = {};
  return {
    baiduAnalytics: t,
    registerBaiduAnalytics: y({
      analytics: t,
      platform: "baidu"
    })
  };
}
function m() {
  const t = {};
  return {
    cnzzAnalytics: t,
    registerCnzzAnalytics: y({
      analytics: t,
      platform: "cnzz"
    })
  };
}
function b({
  analytics: t,
  platform: e
}) {
  return {
    install: (r, { router: n, ...s }) => {
      if (!l)
        return;
      const o = new p({ platform: e, ...s });
      for (const u in o)
        Object.prototype.hasOwnProperty.call(o, u) && (t[u] = o[u]);
      const a = Object.getPrototypeOf(o);
      Object.setPrototypeOf(t, a);
      const i = h(e);
      if (i)
        switch (d(r)) {
          case 2: {
            r.prototype[i] = t;
            break;
          }
          case 3: {
            r.config.globalProperties[i] = t;
            break;
          }
        }
      n && n.afterEach(() => {
        t.trackPageview({
          pageUrl: window.location.href
        });
      });
    }
  };
}
function v() {
  const t = {};
  return {
    baiduAnalytics: t,
    registerBaiduAnalytics: b({
      analytics: t,
      platform: "baidu"
    })
  };
}
function E() {
  const t = {};
  return {
    cnzzAnalytics: t,
    registerCnzzAnalytics: b({
      analytics: t,
      platform: "cnzz"
    })
  };
}
export {
  O as createVitePressBaiduAnalytics,
  m as createVitePressCnzzAnalytics,
  v as createVueBaiduAnalytics,
  E as createVueCnzzAnalytics
};
