// @web-analytics/core/lib/index.mjs
/**
 * name: @web-analytics/core
 * version: v0.1.5
 * description: Website pageview analytics tool for framework-free and multi-analytics-platform support.
 * author: chengpeiquan <chengpeiquan@chengpeiquan.com>
 * homepage: https://analytics.js.org/core/
 * license: MIT
 */
var m = Object.defineProperty;
var y = (t, e, s) => e in t ? m(t, e, { enumerable: !0, configurable: !0, writable: !0, value: s }) : t[e] = s;
var o = (t, e, s) => (y(t, typeof e != "symbol" ? e + "" : e, s), s);
import { pascalCase as P, isBrowser as _, loadRes as z } from "@bassist/utils";
const A = "web-analytics-core", E = [
  /**
   * Baidu analysis platform
   * @website https://tongji.baidu.com
   * @docs https://tongji.baidu.com/open/api
   */
  "baidu",
  /**
   * U-Web(CNZZ) analysis platform
   * @website https://www.umeng.com/web
   * @docs https://developer.umeng.com/docs/67963/detail/74517
   */
  "cnzz"
];
var u = /* @__PURE__ */ ((t) => (t.SetAutoPageview = "_setAutoPageview", t.SetAccount = "_setAccount", t.TrackPageview = "_trackPageview", t.TrackEvent = "_trackEvent", t))(u || {});
function D({
  pluginId: t,
  platform: e,
  websiteId: s,
  propertyKey: r,
  args: n
}) {
  const i = [];
  switch (r) {
    case "loadSdk": {
      i.push("JS-SDK load done.");
      break;
    }
    case "setAccount": {
      i.push("set account done.");
      break;
    }
    case "trackPageview": {
      const { pageUrl: c, fromUrl: h } = n[0] || {};
      i.push("track pageview done."), i.push(`pageUrl:      ${p(c)}`), e === "cnzz" && h && i.push(`fromUrl:      ${p(h)}`);
      break;
    }
    case "trackEvent": {
      const { category: c, action: h, label: f, value: I, nodeId: $ } = n[0] || {};
      i.push("track event done."), i.push(`category:     ${c}`), i.push(`action:       ${h}`), i.push(`label:        ${l(f)}`), i.push(`value:        ${g(I)}`), e === "cnzz" && i.push(`nodeId:       ${b($)}`);
      break;
    }
  }
  i.push(`websiteId:    ${s}`), i.push(`time:         ${/* @__PURE__ */ new Date()}`);
  const a = i.join(`
`);
  return `
[${t}] ${P(e)} Analytics ${a}

`;
}
function p(t) {
  if ((!t || typeof t != "string") && (t = "/"), t.startsWith("http")) {
    const e = t.split("/"), s = `${e[0]}//${e[2]}`;
    t = t.replace(s, "");
  }
  return t;
}
function O(t) {
  return (!t || t && typeof t != "string") && (t = ""), typeof t == "string" && !t.includes("http") && (t = ""), t;
}
function b(t) {
  return (!t || typeof t != "string") && (t = ""), t;
}
function l(t) {
  return (!t || typeof t != "string") && (t = ""), t;
}
function g(t) {
  return (!t || !Number(t)) && (t = 1), t;
}
function w(t, e, s) {
  const r = s.value;
  return s.value = function(...n) {
    const {
      debug: i,
      pluginId: a,
      platform: c,
      websiteId: h
    } = this;
    if (i) {
      const f = D({
        pluginId: a,
        platform: c,
        websiteId: h,
        propertyKey: e,
        args: n
      });
      console.log(f);
    }
    r.apply(this, n);
  }, s;
}
function S(t, e, s) {
  const r = s.value;
  return s.value = function(...n) {
    const [i] = n, { category: a, action: c } = i;
    if (typeof a != "string" || typeof c != "string" || !a || !c) {
      this.throwError(
        "Valid `category` and `action` are missing from the track event options."
      );
      return;
    }
    r.apply(this, n);
  }, s;
}
var T = Object.defineProperty, j = Object.getOwnPropertyDescriptor, L = (t, e, s, r) => {
  for (var n = r > 1 ? void 0 : r ? j(e, s) : e, i = t.length - 1, a; i >= 0; i--)
    (a = t[i]) && (n = (r ? a(e, s, n) : a(n)) || n);
  return r && n && T(e, s, n), n;
};
class k {
  constructor({
    pluginId: e,
    platform: s,
    websiteId: r,
    debug: n
  }) {
    o(this, "pluginId");
    o(this, "platform");
    o(this, "sdkInstance");
    o(this, "sdkUrl", "");
    o(this, "websiteId", "");
    o(this, "debug");
    this.pluginId = e || A, this.platform = s, this.websiteId = r, this.debug = n ?? !1, this.updatePlatformInfo(), this.loadSdk();
  }
  /**
   * Smooth out the differences between different platforms
   */
  updatePlatformInfo() {
    if (_)
      switch (this.platform) {
        case "baidu": {
          this.sdkInstance = window._hmt || [], this.sdkUrl = `https://hm.baidu.com/hm.js?${this.websiteId}`;
          break;
        }
        case "cnzz": {
          this.sdkInstance = window._czc || [], this.sdkUrl = `https://v1.cnzz.com/z_stat.php?id=${this.websiteId}&web_id=${this.websiteId}`;
          break;
        }
        default: {
          const e = E.join(", ");
          this.throwError(
            `Unsupported platform options, only supported: ${e}.`
          );
        }
      }
  }
  loadSdk() {
    !this.sdkInstance || !this.sdkUrl || (this.sdkInstance.push([u.SetAutoPageview, !1]), z({
      type: "js",
      id: `${this.pluginId}-${this.platform}-${this.websiteId}`,
      resource: this.sdkUrl
    }).catch((e) => {
      console.log(e);
    }));
  }
  throwError(e) {
    throw new Error(`[${this.pluginId}] ${e}`);
  }
}
L([
  w
], k.prototype, "loadSdk", 1);
var C = Object.defineProperty, F = Object.getOwnPropertyDescriptor, v = (t, e, s, r) => {
  for (var n = r > 1 ? void 0 : r ? F(e, s) : e, i = t.length - 1, a; i >= 0; i--)
    (a = t[i]) && (n = (r ? a(e, s, n) : a(n)) || n);
  return r && n && C(e, s, n), n;
};
class d extends k {
  constructor({
    pluginId: e,
    platform: s,
    websiteId: r,
    debug: n
  }) {
    super({ pluginId: e, platform: s, websiteId: r, debug: n });
  }
  /**
   * Provide multi-account switching for upper-level plugins
   */
  setAccount() {
    this.sdkInstance && this.sdkInstance.push([u.SetAccount, this.websiteId]);
  }
  trackPageview({
    pageUrl: e,
    // @ts-ignore
    fromUrl: s
  }) {
    if (this.sdkInstance)
      switch (this.setAccount(), this.platform) {
        case "baidu": {
          this.sdkInstance.push([
            u.TrackPageview,
            p(e)
          ]);
          break;
        }
        case "cnzz": {
          this.sdkInstance.push([
            u.TrackPageview,
            p(e),
            s ? O(s) : ""
          ]);
          break;
        }
      }
  }
  trackEvent({
    category: e,
    action: s,
    label: r,
    value: n,
    // @ts-ignore
    nodeId: i
  }) {
    if (this.sdkInstance)
      switch (this.setAccount(), this.platform) {
        case "baidu": {
          this.sdkInstance.push([
            u.TrackEvent,
            e,
            s,
            l(r),
            g(n)
          ]);
          break;
        }
        case "cnzz": {
          this.sdkInstance.push([
            u.TrackEvent,
            e,
            s,
            l(r),
            g(n),
            b(i)
          ]);
          break;
        }
      }
  }
}
v([
  w
], d.prototype, "trackPageview", 1);
v([
  w,
  S
], d.prototype, "trackEvent", 1);
function M(t) {
  return new d({
    platform: "baidu",
    ...t
  });
}
function R(t) {
  return new d({
    platform: "cnzz",
    ...t
  });
}
export {
  d as Analytics,
  E as SUPPORTED_ANALYTICS_PLATFORMS,
  M as createBaiduAnalytics,
  R as createCnzzAnalytics
};
